internal/state/ownership/tree.go

```diff
@@ -0,0 +1,180 @@
+	}
+
+	t, err := sparsemt.New(TreeDepth, store, contract.String()+TreePrefix)
+	if err != nil {
+		return nil, err
+	}
+
+	root, err := headRoot(contract, store)
+	if err != nil {
+		return nil, err
+	}
+
+	t.SetRoot(root)
+	slog.Debug("enumeratedTokenTree", "HEAD", root.String())
+
+	return &tree{contract, t, store, false}, err
+}
+
+// TokenTransfer updates the SlotOwner of the token
+func (b *tree) Transfer(eventTransfer scan.EventTransfer) error {
+
+	tokenData, err := b.TokenData(eventTransfer.TokenId)
+	if err != nil {
+		return err
+	}
+
+	tokenData.SlotOwner = eventTransfer.To
+	if err := b.SetTokenData(tokenData, eventTransfer.TokenId); err != nil {
+		return err
+	}
+	return setHeadRoot(b.contract, b.store, b.Root())
+}
+
+// Mint creates a new token
+func (b *tree) Mint(tokenId *big.Int, idx int) error {
+	tokenData, err := b.TokenData(tokenId)
+	if err != nil {
+		return err
+	}
+
+	tokenData.Minted = true
+	tokenData.Idx = idx
+	if err := b.SetTokenData(tokenData, tokenId); err != nil {
+		return err
+	}
+
+	return setHeadRoot(b.contract, b.store, b.Root())
+}
+
+// SetTokenData updates the tokenData
+func (b *tree) SetTokenData(tokenData *TokenData, tokenId *big.Int) error {
+	buf, err := json.Marshal(tokenData)
+	if err != nil {
+		return err
+	}
+
+	hash := crypto.Keccak256Hash(buf)
+	if err := b.store.Set([]byte(b.contract.String()+TokenDataPrefix+hash.String()), buf); err != nil {
+		return err
+	}
+
+	return b.mt.SetLeaf(tokenId, hash)
+}
+
+// TokenData returns the tokenData
+func (b *tree) TokenData(tokenId *big.Int) (*TokenData, error) {
+	leaf, err := b.mt.Leaf(tokenId)
+	if err != nil {
+		return &TokenData{common.Address{}, false, 0}, err
+	}
+	if leaf.String() == sparsemt.Null {
+		SlotOwner := initSlotOwner(tokenId)
+		return &TokenData{SlotOwner, false, 0}, nil
+	}
+
+	buf, err := b.store.Get([]byte(b.contract.String() + TokenDataPrefix + leaf.String()))
+	if err != nil {
+		return &TokenData{common.Address{}, false, 0}, err
+	}
+
+	var tokenData TokenData
+	if err := json.Unmarshal(buf, &tokenData); err != nil {
+		return &TokenData{common.Address{}, false, 0}, err
+	}
+
+	return &tokenData, nil
+}
+
+func (b *tree) OwnerOf(tokenId *big.Int) (common.Address, error) {
+	tokenData, err := b.TokenData(tokenId)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	if tokenData.Minted {
+		return tokenData.SlotOwner, nil
+	}
+
+	return common.Address{}, nil
+}
+
+func initSlotOwner(tokenId *big.Int) common.Address {
+	SlotOwner, _ := new(big.Int).SetString(Ones160bits, 16) // omit success. this is constant and it will always be true
+	return common.BigToAddress(SlotOwner.And(SlotOwner, tokenId))
+}
+
+func (b *tree) Root() common.Hash {
+	return b.mt.Root()
+}
+
+func headRoot(contract common.Address, store storage.Tx) (common.Hash, error) {
+	buf, err := store.Get([]byte(contract.String() + HeadRootKey))
+	if err != nil {
+		return common.Hash{}, err
```

Comment:
`store.Get` might return `ErrKeyNotFound`, and in this case you should return `common.Hash{}, nil`.
We created the accessory method `get` in the `repository` module so the caller doesn't have to handle this situation everytime.
Source: https://github.com/freeverseio/laos-universal-node/blob/main/internal/repository/repository.go#L60

Maybe we can integrate the repository inside each tree. Let's discuss this on Monday

Reply:
we mentioned this in private chat. Better to put this check in storage interface in my opinion